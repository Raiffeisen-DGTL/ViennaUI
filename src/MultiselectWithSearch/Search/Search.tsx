import React, { FormEvent, forwardRef, KeyboardEvent, Ref, useEffect, useRef, useState } from 'react';
import { Box, Input, Helper } from './Search.styles';
import { defer } from '../../Utils';

export interface SearchProps {
    value?: string;
    placeholder?: string;
    disabled?: boolean;
    inputRef?: Ref<HTMLInputElement>;
    onChange?: (value: string) => void;
    onFocus?: () => void;
    onBlur?: () => void;
    onDelete?: (event: KeyboardEvent<HTMLInputElement>) => void;
    onChangeWidth?: () => void;
    testId?: string;
}

export const Search = forwardRef<HTMLDivElement, SearchProps>(
    ({ value, placeholder, disabled, inputRef, onChange, onFocus, onBlur, onDelete, onChangeWidth, testId }, ref) => {
        const [state, setState] = useState(value ?? '');
        const [width, setWidth] = useState(10);
        const textRef = useRef<HTMLSpanElement>(null);

        const handleChange = (event: FormEvent<HTMLInputElement>) => {
            const str = event.currentTarget.value;
            setState(str);
            defer(() => {
                onChange && onChange(str);
            });
        };

        const handleKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
            if (onDelete && state.length === 0 && event.key === 'Backspace' && !event.repeat) {
                onDelete(event);
            }
        };

        const handleFocus = () => {
            onFocus && onFocus();
        };

        useEffect(() => {
            setState(value ?? '');
        }, [value]);

        useEffect(() => {
            setWidth((textRef.current?.offsetWidth ?? 0) + 10);
        }, [state]);

        useEffect(() => {
            defer(() => {
                onChangeWidth && onChangeWidth();
            });
        }, [width]);

        return (
            <Box ref={ref} $width={width}>
                <Input
                    ref={inputRef}
                    value={state}
                    placeholder={placeholder}
                    type='text'
                    disabled={disabled}
                    onInput={handleChange}
                    onKeyDown={handleKeyDown}
                    onFocus={handleFocus}
                    onBlur={onBlur}
                    data-test-id={testId}
                    data-testid={testId}
                />
                <Helper ref={textRef}>{state}</Helper>
            </Box>
        );
    }
);
