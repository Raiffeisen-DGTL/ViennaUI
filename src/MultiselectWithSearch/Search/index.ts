export { Search, type SearchProps } from './Search';
