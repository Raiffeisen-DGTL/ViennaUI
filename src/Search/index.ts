export { Search, type SearchProps, type SearchEvent, type SearchSelectEvent, type ChildrenFunc } from './Search';
