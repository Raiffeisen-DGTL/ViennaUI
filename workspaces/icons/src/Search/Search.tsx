import React, { SVGAttributes } from 'react';

export interface SearchProps extends SVGAttributes<SVGElement> {
    [key: string]: any;
    size?: 'xs' | 's' | 'm' | 'l' | 'xl' | number;
    color?: string;
}

const sizes = { xs: 12, s: 16, m: 20, l: 24, xl: 32 };
export const Search: React.FC<SearchProps> = (props): JSX.Element => {
    const { color = 'currentColor', size = 'm', ...attrs } = props;
    const d = sizes[size] || size;

    return (
        <svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24' width={d} height={d} fill={color} {...attrs}>
            <defs />
            <path d='M11 3a8 8 0 016.32 12.905l3.887 3.888-1.414 1.414-3.888-3.886A8 8 0 1111 3zm0 2a6 6 0 100 12 6 6 0 000-12z' />
        </svg>
    );
};

Search.defaultProps = {
    size: 'm',
    color: 'currentColor',
};

Search.displayName = 'Search';
export default Search;
