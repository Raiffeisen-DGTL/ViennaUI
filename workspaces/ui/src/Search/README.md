# Search & Suggest

Компонент поиска данных с отображением подстовляемого значения.

Является оберткой над компонентом `Input` и наследует все его свойства, такие как: дизайн, размер, состояния и т.д.

## Props

| Prop | Type | Default | Description |
| --- | --- | --- | --- |
| ref | ((instance: HTMLInputElement \| null) => void) \| RefObject \| null \| undefined | false | Сcылка на нативный элемент input, доступна после отрисовки |
| size | "xs" \| "s" \| "l" \| "xl" \| undefined | false | Размеры |
| design | "outline" \| "material" \| undefined | false | Дизайн |
| prefix | ReactNode | false | Значанеие отображаемое перед компонентом |
| postfix | ReactNode | false | Значение отображаемое после компонента |
| suggests | any[] \| undefined | false | Список элементов в выпадающем списке: массив |
| value | any | false | Выбранный элемент (должен совпадать по интерфейсу с объектами массива) |
| maxListHeight | number \| undefined | false | Максимальная высота выпадающего списка в пикселях |
| invalid | boolean \| undefined | false | Компонент отображается как ошибочный если true |
| children | string \| number \| boolean \| {} \| ReactElement ReactElement Component)> \| null) \| (new (props: any) => Component)> \| ... 4 more ... \| undefined | false | Элементы выпадающего списка в случае если не используется свойство suggests |
| onSelect | SearchSelectEvent \| undefined | false | Обработчик события при выборе элемента списка |
| onChange | SearchEvent> \| undefined | false | Обработчик события при наборе текста в поле ввода |
| onFocus | InputEvent> \| undefined | false | Обработчик события при получении фокуса компонентом |
| onBlur | InputEvent> \| undefined | false | Обработчик события при потере фокуса компонентом |
| onKeyDown | SearchEvent<(event: KeyboardEvent) => void> \| undefined | false | Обработчик события при нажатии кнопки клавиатуры, когда компонент в фокусе |
| onKeyUp | ((event: KeyboardEvent) => void) \| undefined | false | Обработчик события при отпускании кнопки клавиатуры, когда компонент в фокусе |
| onKeyPress | ((event: KeyboardEvent) => void) \| undefined | false | Обработчик события при нажатии и удержании кнопки клавиатуры с печатемым символом, когда компонент в фокусе |
| onScroll | ((event: UIEvent) => void) \| undefined | false | Обработчик события при прокрутке списка |
| showInlineSuggest | boolean \| undefined | false | Показывать строчное дополнение (по умолчанию true) |
| wrapSuggestions | boolean \| undefined | false | Разрешить перенос подсказки на следующую строку |
| valueToString | ((item?: any) => string) \| undefined | false | Определяем значение которое надо вывести в компонент как текст выбранного значения |

```jsx
const [state, setValue] = React.useState({ value: '', suggests: [] });
const handleSelect = (e, data) => setValue({ ...state, value: data.value });
const handleChange = (e, data) => {
    const mock = [
        'Райффайзен банк',
        'Райффайзен IT',
        'Иванов Иван Иванович',
        'Иванов Иван Петрович',
        'Ивановская область',
        'Рамблер',
    ];
    const regexp = new RegExp(`^${escape(data.value.toUpperCase())}`);
    const temp = (data.value && mock.filter((m) => regexp.test(escape(m.toUpperCase())))) || [];
    setValue({ value: data.value, suggests: temp });
};
return (
    <Search
        suggests={state.suggests}
        value={state.value}
        placeholder='Начните ввод'
        onChange={handleChange}
        onSelect={handleSelect}
    />
);
```

## Использование

Начните набирать: `Райффайзен` или `Иванов`. Выбор по `Enter` или щелчком по элементу списка.

```jsx
{
    () => {
        const [value, setValue] = React.useState();
        const [suggests, setSuggests] = React.useState();
        const handleChange = React.useCallback((e, data) => {
            const mock = [
                'Райффайзен банк',
                'Райффайзен IT',
                'Иванов Иван Иванович',
                'Иванов Иван Петрович',
                'Ивановская область',
                'Рамблер',
            ];
            const regexp = new RegExp(`^${escape(data.value.toUpperCase())}`);
            const temp = (data.value && mock.filter((m) => regexp.test(escape(m.toUpperCase())))) || [];
            setSuggests(temp);
            setValue(data.value);
        }, []);
        const handleSelect = React.useCallback((e, data) => {
            setValue(data.value);
        }, []);
        return (
            <Groups design='vertical'>
                <Search
                    suggests={suggests}
                    value={value}
                    placeholder={'Начните ввод'}
                    onChange={handleChange}
                    onSelect={handleSelect}
                />
            </Groups>
        );
    };
}
```

## Кастомизация

#### Префиксная иконка

```jsx
const [value, setValue] = React.useState();
const [suggests, setSuggests] = React.useState();
const handleChange = React.useCallback((e, data) => {
    const mock = [
        'Райффайзен банк',
        'Райффайзен IT',
        'Иванов Иван Иванович',
        'Иванов Иван Петрович',
        'Ивановская область',
        'Рамблер',
    ];
    const regexp = new RegExp(`^${data.value.toUpperCase()}`);
    const temp = (data.value && mock.filter((m) => regexp.test(m.toUpperCase()))) || [];
    setSuggests(temp);
    setValue(data.value);
}, []);
const handleSelect = React.useCallback((e, data) => {
    setValue(data.value);
}, []);
return (
    <Search
        prefix={<Violin />}
        suggests={suggests}
        value={value}
        placeholder={'Начните ввод'}
        onChange={handleChange}
        onSelect={handleSelect}
    />
);
```

#### Постфиксная иконка

```jsx
const [value, setValue] = React.useState();
const [suggests, setSuggests] = React.useState();
const handleChange = React.useCallback((e, data) => {
    const mock = [
        'Райффайзен банк',
        'Райффайзен IT',
        'Иванов Иван Иванович',
        'Иванов Иван Петрович',
        'Ивановская область',
        'Рамблер',
    ];
    const regexp = new RegExp(`^${data.value.toUpperCase()}`);
    const temp = (data.value && mock.filter((m) => regexp.test(m.toUpperCase()))) || [];
    setSuggests(temp);
    setValue(data.value);
}, []);
const handleSelect = React.useCallback((e, data) => {
    setValue(data.value);
}, []);
return (
    <Search
        postfix={<Theater2 />}
        suggests={suggests}
        value={value}
        placeholder={'Начните ввод'}
        onChange={handleChange}
        onSelect={handleSelect}
    />
);
```

#### Две иконки

```jsx
const [value, setValue] = React.useState();
const [suggests, setSuggests] = React.useState();
const handleChange = React.useCallback((e, data) => {
    const mock = [
        'Райффайзен банк',
        'Райффайзен IT',
        'Иванов Иван Иванович',
        'Иванов Иван Петрович',
        'Ивановская область',
        'Рамблер',
    ];
    const regexp = new RegExp(`^${data.value.toUpperCase()}`);
    const temp = (data.value && mock.filter((m) => regexp.test(m.toUpperCase()))) || [];
    setSuggests(temp);
    setValue(data.value);
}, []);
const handleSelect = React.useCallback((e, data) => {
    setValue(data.value);
}, []);
return (
    <Search
        prefix={<Task />}
        postfix={<Spinner />}
        suggests={suggests}
        value={value}
        placeholder={'Начните ввод'}
        onChange={handleChange}
        onSelect={handleSelect}
    />
);
```

## Кастомизация элементов списка

В качестве `дочернего элемента` можно передать только `функцию`. Функция принемает аргумент объект c полями: `props, suggest, value` где:

-   props - прокинутые насквозь свойства Search
-   suggest - текст текущего элеменнта
-   value - значение ввода

Возвращать следует валидный `JSX.Element`, который станет шаблоном для всех элементов

#### Изменение иконки

```jsx
const [value, setValue] = React.useState();
const [suggests, setSuggests] = React.useState();
const handleChange = React.useCallback((e, data) => {
    const mock = [
        'Райффайзен банк',
        'Райффайзен IT',
        'Иванов Иван Иванович',
        'Иванов Иван Петрович',
        'Ивановская область',
        'Рамблер',
    ];
    const regexp = new RegExp(`^${data.value.toUpperCase()}`);
    const temp = (data.value && mock.filter((m) => regexp.test(m.toUpperCase()))) || [];
    setSuggests(temp);
    setValue(data.value);
}, []);
const handleSelect = React.useCallback((e, data) => {
    setValue(data.value);
}, []);
return (
    <Search
        prefix={<Task />}
        postfix={<Spinner />}
        suggests={suggests}
        value={value}
        placeholder={'Начните ввод'}
        onChange={handleChange}
        onSelect={handleSelect}>
        {({ props, suggest, value }) => <CustomItemWithIcon icon={<OperationsHistory />} value={suggest} />}
    </Search>
);
```

#### Пример подсветки совпадений

```jsx
const [value, setValue] = React.useState();
const [suggests, setSuggests] = React.useState();
const handleChange = React.useCallback((e, data) => {
    const mock = [
        'Райффайзен банк',
        'Райффайзен IT',
        'Иванов Иван Иванович',
        'Иванов Иван Петрович',
        'Ивановская область',
        'Рамблер',
    ];
    const regexp = new RegExp(`^${data.value.toUpperCase()}`);
    const temp = (data.value && mock.filter((m) => regexp.test(m.toUpperCase()))) || [];
    setSuggests(temp);
    setValue(data.value);
}, []);
const handleSelect = React.useCallback((e, data) => {
    setValue(data.value);
}, []);
return (
    <Search
        prefix={<Task />}
        postfix={<Spinner />}
        suggests={suggests}
        value={value}
        placeholder={'Начните ввод'}
        onChange={handleChange}
        onSelect={handleSelect}>
        {({ props, suggest, value }) => <CustomItem value={suggest} match={value} />}
    </Search>
);
```

#### Работа с объектами

Если вы передаете в качестве `value` произвольный объект то `должны` определить функцию `valueToString` обязательно возвращающую строку.

-   Объекты `suggests` должны быть одной структуры с `value`.
-   Возвращаемое значение в `onChange` - строка в инпуте
-   Возвращаемое значение в `onSelect` - объек

Возможно вам захочется отключить строковую подсказку, для этого установите свойство `showInlineSuggest` в значение `false`.

```jsx
const [value, setValue] = React.useState();
const [suggests, setSuggests] = React.useState();
const valueToString = React.useCallback((val) => {
    if (val) {
        if (typeof val === 'string') {
            return val;
        }
        return `${val.family} ${val.name} ${val.middle}`;
    }
    return '';
}, []);
const handleChange = React.useCallback((e, data) => {
    const mock = [
        { family: 'Иванов', name: 'Иван', middle: 'Иванович' },
        { family: 'Иванов', name: 'Иван', middle: 'Петрович' },
        { family: 'Иванов', name: 'Константин', middle: 'Петрович' },
    ];
    const regexp = new RegExp(`^${escape(data.value.toUpperCase())}`);
    const temp = (data.value && mock.filter((m) => regexp.test(escape(valueToString(m).toUpperCase())))) || [];
    setSuggests(temp);
    setValue(data.value);
}, []);
const handleSelect = React.useCallback((e, data) => {
    setValue(data.value);
}, []);
return (
    <Search
        prefix={<Task />}
        postfix={<Spinner />}
        showInlineSuggest={false}
        suggests={suggests}
        value={value}
        placeholder={'Начните ввод'}
        valueToString={valueToString}
        onChange={handleChange}
        onSelect={handleSelect}>
        {({ props, suggest, value }) => <CustomItem value={valueToString(suggest)} match={valueToString(value)} />}
    </Search>
);
```
