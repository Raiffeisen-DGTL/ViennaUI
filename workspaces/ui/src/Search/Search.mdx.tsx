import React from 'react';
import styled from 'styled-components';

const Box = styled.div`
    display: flex;
    flex-direction: column;
`;

const Value = styled.div`
    display: flex;
    color: black;
`;

const Box2 = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;

    svg {
        min-width: 32px;
        margin-right: 8px;
    }
`;

const Highlighter = styled.div`
    display: inline-block;
    background-color: #fee600;
`;

export const CustomItem = (props) => {
    const { value, match } = props;

    const computed = value
        .replace(/ /gi, '\u00a0')
        .split(new RegExp(`${match.replace(/ /gi, '\u00a0')}`, 'gm'))
        .reduce(
            (o, i, idx, a) =>
                idx < a.length - 1
                    ? [...o, i, <Highlighter key={idx}>{match.replace(/ /gi, '\u00a0')}</Highlighter>]
                    : [...o, i],
            []
        );

    return (
        <Box>
            <Value>{computed}</Value>
        </Box>
    );
};

export const CustomItemWithIcon = (props) => {
    return (
        <Box2>
            {props.icon}
            <Box>
                <Value>{props.value}</Value>
            </Box>
        </Box2>
    );
};

export const Warn = styled.div`
    color: black;
    background-color: #FFCCBB;
    padding: 12px 10px;
    box-sizing: border-box;
    border-radius: 4px;
    display: flex;
    align-items: center;

    svg:first-child{
        min-width: 32px;
        margin-right: 12px;
    }

    span{
        font-family: Inconsolata,Consolas,monospace;
        margin: 0;
        padding: 4px 5px;
        border-radius: 4px;
        white-space: nowrap;
        background-color: #FF9B78
    }
}
`;

export const Info = styled.div`
    color: black;
    background-color: #B3D4D8;
    padding: 12px 10px;
    box-sizing: border-box;
    border-radius: 4px;
    display: flex;
    align-items: center;
    margin-top: 8px;

    svg:first-child{
        min-width: 32px;
        margin-right: 12px;
    }

    span{
        font-family: Inconsolata,Consolas,monospace;
        margin: 0;
        padding: 4px 5px;
        border-radius: 4px;
        white-space: nowrap;
        background-color: #45A3B2
    }
}
`;
