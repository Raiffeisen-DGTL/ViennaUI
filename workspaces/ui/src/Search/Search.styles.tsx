import styled from 'styled-components';
import { getPresets } from 'vienna.ui-primitives';

const search = getPresets('search', {
    base: null,
    custom: null,
});

export const Box = styled.div`
    position: relative;
    ${search.base}
    ${search.custom}
`;
