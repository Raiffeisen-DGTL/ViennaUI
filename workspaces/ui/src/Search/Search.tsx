/* eslint-disable react-hooks/rules-of-hooks */
import React, { useCallback, useState, useRef, RefAttributes, ForwardRefExoticComponent } from 'react';
import { Box } from './Search.styles';
import { Input, InputEvent } from '../Input';
import { DropList } from '../DropList';
import { Item, Props as ItemProps } from '../DropList/Item';

interface Data {
    props: any;
    suggest: string;
    value: any;
}
export type ChildrenFunc = (data: Data) => React.ReactNode;

export type SearchEvent<T = React.FormEvent<HTMLInputElement>> = (
    event: T,
    data?: { name?: string; value?: string; index: number }
) => void;

export type SearchSelectEvent = (
    event: React.FormEvent<HTMLInputElement>,
    data?: { name?: string; value: any }
) => void;

export interface SearchProps {
    [key: string]: any;
    /** Сcылка на нативный элемент input, доступна после отрисовки */
    ref?: React.Ref<HTMLInputElement>;

    /** Размеры */
    size?: 'xs' | 's' | 'l' | 'xl';

    /** Дизайн */
    design?: 'outline' | 'material';

    /** Значанеие отображаемое перед компонентом  */
    prefix?: React.ReactNode;

    /** Значение отображаемое после компонента  */
    postfix?: React.ReactNode;

    /** ID компонента */
    id?: string;
    /** Имя компонента */
    name?: string;

    /** Список элементов в выпадающем списке: массив */
    suggests?: any[];

    /** Выбранный элемент (должен совпадать по интерфейсу с объектами массива) */
    value?: any;

    /** Максимальная высота выпадающего списка в пикселях */
    maxListHeight?: number;
    /** Максимальная ширина выпадающего списка в пикселях */
    maxListWidth?: number;

    /** Компонент неактивен если true */
    disabled?: boolean;

    /** Компонент отображается как ошибочный если true */
    invalid?: boolean;

    /** Элементы выпадающего списка в случае если не используется свойство suggests  */
    children?: ChildrenFunc;

    /** Обработчик события при выборе элемента списка  */
    onSelect?: SearchSelectEvent;

    /** Обработчик события при наборе текста в поле ввода  */
    onChange?: SearchEvent;

    /** Обработчик события при получении фокуса компонентом  */
    onFocus?: InputEvent<React.FormEvent<HTMLInputElement>>;

    /** Обработчик события при потере фокуса компонентом  */
    onBlur?: InputEvent<React.FormEvent<HTMLInputElement>>;

    /** Обработчик события при нажатии кнопки клавиатуры, когда компонент в фокусе  */
    onKeyDown?: SearchEvent<React.KeyboardEventHandler<HTMLInputElement>>;

    /** Обработчик события при отпускании кнопки клавиатуры, когда компонент в фокусе  */
    onKeyUp?: React.KeyboardEventHandler<HTMLInputElement>;

    /** Обработчик события при нажатии и удержании кнопки клавиатуры с печатемым символом, когда компонент в фокусе  */
    onKeyPress?: React.KeyboardEventHandler<HTMLInputElement>;

    /** Обработчик события при прокрутке списка  */
    onScroll?: React.UIEventHandler<HTMLDivElement>;

    /** Показывать строчное дополнение (по умолчанию true)  */
    showInlineSuggest?: boolean;

    /** Текст плейсхолдера */
    placeholder?: string;
    /** Если установлено то выравнивает содержимое списка по ширине компонента (по умолчанию true)  */
    fitOptions?: boolean;

    /** Разрешить перенос подсказки на следующую строку */
    wrapSuggestions?: boolean;

    /** Определяем значение которое надо вывести в компонент как текст выбранного значения */
    valueToString?: (item?: any) => string;
    fixed?: boolean;
}

const handleMouseDownItem = (event) => event.preventDefault();

export const Search: React.FC<SearchProps> & { Item: React.FC<ItemProps> } = React.forwardRef(
    (props: SearchProps, ref: React.Ref<HTMLInputElement>) => {
        const {
            id,
            name,
            disabled,
            value,
            suggests = [],
            onSelect,
            onChange,
            children,
            design,
            onBlur,
            onFocus,
            onKeyDown,
            size,
            onScroll,
            valueToString = (val): string => val?.toString() || '',
            maxListHeight,
            maxListWidth,
            showInlineSuggest,
            wrapSuggestions,
            fitOptions = true,
            fixed,
            ...attrs
        } = props;

        if (children && typeof children !== 'function') {
            console.error('В качестве children может быть только функция'); // eslint-disable-line
            return <></>;
        }

        const inputRef = useCallback((input) => {
            if (input) {
                if (typeof ref === 'function') {
                    (ref as any)(input);
                }
                if (ref && typeof ref === 'object') {
                    (ref as any).current = input;
                }
                input.style.background = 'transparent';
                input.style.position = 'relative';
            }
        }, []);

        const [currentIndex, setCurrentIndex] = useState(0);
        const [active, setActive] = useState(false);

        const searchElementRef = useRef<HTMLDivElement>(null);

        const handleBlur = useCallback(
            (e, data): void => {
                setActive(false);
                if (onBlur) {
                    onBlur(e, data);
                }
            },
            [onBlur]
        );

        const handleFocus = useCallback(
            (e, data): void => {
                e.stopPropagation();
                setActive(true);
                if (onFocus) {
                    onFocus(e, data);
                }
            },
            [onFocus]
        );

        const handleChange = useCallback(
            (e, data): void => {
                if (!disabled && onChange) {
                    setActive(true);
                    onChange(e, { ...data, index: currentIndex });
                }
            },
            [onChange, disabled, currentIndex]
        );

        const handleSelect = useCallback(
            (e, value): void => {
                if (!disabled && onSelect) {
                    if (onChange) {
                        onChange(e, { name, value: valueToString(value) || '', index: currentIndex });
                    }

                    onSelect(e, { name, value });
                    setActive(false);
                }
            },
            [onSelect, name, disabled, currentIndex, value]
        );

        const handleHide = useCallback(() => {
            setActive(false);
        }, []);

        const hadleKeyDown = useCallback(
            (e): void => {
                const lastIndex = suggests.length - 1;

                switch (e.key) {
                    case 'ArrowDown': {
                        e.preventDefault();
                        const computed = currentIndex < lastIndex ? currentIndex + 1 : 0;
                        setCurrentIndex(computed);
                        if (onKeyDown) {
                            onKeyDown(e, { name, value, index: computed });
                        }
                        break;
                    }
                    case 'ArrowUp': {
                        e.preventDefault();
                        const computed = currentIndex > 0 ? currentIndex - 1 : lastIndex;
                        setCurrentIndex(computed);
                        if (onKeyDown) {
                            onKeyDown(e, { name, value, index: computed });
                        }
                        break;
                    }
                    case 'Enter':
                        e.preventDefault();
                        if (suggests[currentIndex]) {
                            handleSelect(e, suggests[currentIndex]);
                        }
                        if (onKeyDown) {
                            onKeyDown(e, { name, value, index: currentIndex });
                        }
                        break;
                    default:
                        if (onKeyDown) {
                            onKeyDown(e, { name, value, index: currentIndex });
                        }
                        break;
                }
            },
            [onKeyDown, name, currentIndex, value, suggests, handleSelect]
        );

        const showList = active && !disabled && Boolean(suggests.length);

        const constructItems = useCallback(() => {
            return suggests.map((suggest, index) => {
                const handleSelectItem = (e) => handleSelect(e, suggest);
                return (
                    <Item
                        key={index}
                        value={suggest}
                        selected={index === currentIndex}
                        wrapLine={wrapSuggestions}
                        onMouseDown={handleMouseDownItem}
                        onClick={handleSelectItem}>
                        {children ? children({ props, suggest, value }) : valueToString(suggest) || ''}
                    </Item>
                );
            });
        }, [handleSelect, children, currentIndex, props, suggests, value, valueToString, wrapSuggestions]);

        const constructSuggest = useCallback((): string => {
            let suggest = valueToString(suggests[currentIndex] || '');
            if (suggest && value) {
                suggest = suggest.replace(new RegExp(`^.{0,${value.length}}`, 'gmi'), value);
            } // подменяем начало suggest введеным value что бы избежать проблемм с регистром

            return (active && showInlineSuggest && suggest) || ''; // показываем suggest только если поле активно
        }, [showInlineSuggest, suggests, active, currentIndex, valueToString, value]);

        return (
            <Box id={id} ref={searchElementRef}>
                <Input
                    id={id && `ds-${id}-input`}
                    name={name}
                    size={size}
                    design={design}
                    ref={inputRef}
                    value={valueToString(value) || ''}
                    disabled={disabled}
                    smartPlaceholder={constructSuggest()}
                    onChange={handleChange}
                    onKeyDown={hadleKeyDown}
                    onBlur={handleBlur}
                    onFocus={handleFocus}
                    {...attrs}
                />
                {showList && (
                    <DropList
                        maxHeight={maxListHeight}
                        width={maxListWidth}
                        fitItems={fitOptions}
                        fixed={fixed}
                        followParentWhenScroll={fixed}
                        onHide={handleHide}
                        onScroll={onScroll}>
                        {constructItems()}
                    </DropList>
                )}
            </Box>
        );
    }
) as ForwardRefExoticComponent<SearchProps & RefAttributes<HTMLInputElement>> & { Item: React.FC<ItemProps> };

Search.displayName = 'Search';
Search.Item = Item;
Search.defaultProps = {
    design: 'outline',
    size: 'l',
    suggests: [],
    maxListHeight: 300,
    showInlineSuggest: true,
};

export default Search;
