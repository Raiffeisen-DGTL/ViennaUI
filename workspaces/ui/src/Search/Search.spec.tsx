import React from 'react';
import { Search, SearchProps } from './Search';

test('Search /w design', () => {
    const designs: SearchProps['design'][] = ['outline', 'material'];

    designs.forEach((design) => {
        const snap = snapshot.render(<Search design={design} />);
        expect(snap).toMatchSnapshot();
    });
});

test('Search /w size', () => {
    const sizes: SearchProps['size'][] = ['xs', 's', 'l', 'xl'];

    sizes.forEach((size) => {
        const snap = snapshot.render(<Search size={size} />);
        expect(snap).toMatchSnapshot();
    });
});

test('Search /w invalid', () => {
    const snap = snapshot.render(<Search invalid />);
    expect(snap).toMatchSnapshot();
});

test('Search /w onFocus()', () => {
    const fn = jest.fn();
    const snap = snapshot.mount(<Search onFocus={fn} />);

    snap.find('input').simulate('focus');
    expect(fn).toBeCalled();
});

test('Search /w onBlur()', () => {
    const fn = jest.fn();
    const snap = snapshot.mount(<Search onBlur={fn} />);

    snap.find('input').simulate('blur');
    expect(fn).toBeCalled();
});

test('Search w/ any props', () => {
    const props = {
        title: 'title',
    };

    const component = <Search {...props} />;

    const snap = snapshot.render(component);
    expect(snap).toMatchSnapshot();

    const wrapper = snapshot.mount(component);
    for (const [key, value] of Object.entries(props)) {
        expect(wrapper.find('input').props()[key]).toEqual(value);
    }
});
