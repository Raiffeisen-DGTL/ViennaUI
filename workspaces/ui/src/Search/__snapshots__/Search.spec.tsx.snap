// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search /w design 1`] = `
<div
  aria-autocomplete="list"
  aria-disabled="false"
  aria-invalid="false"
  class="Searchstyles__Box-sc-1lyvud3-0 boYLWE"
  role="combobox"
>
  <label
    class="Inputstyles__Wrapper-hc8ttv-6 daAZxl"
  >
    <div
      class="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
    >
      <div
        class="Inputstyles__Box-hc8ttv-1 bzkyqR"
      >
        <input
          aria-invalid="false"
          class="Inputstyles__StyledInput-hc8ttv-2 bUBKOu"
          value=""
        />
      </div>
    </div>
  </label>
</div>
`;

exports[`Search /w design 2`] = `
<div
  aria-autocomplete="list"
  aria-disabled="false"
  aria-invalid="false"
  class="Searchstyles__Box-sc-1lyvud3-0 boYLWE"
  role="combobox"
>
  <label
    class="Inputstyles__Wrapper-hc8ttv-6 lkLMMk"
  >
    <div
      class="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
    >
      <div
        class="Inputstyles__Box-hc8ttv-1 bzkyqR"
      >
        <input
          aria-invalid="false"
          class="Inputstyles__StyledInput-hc8ttv-2 dyFbgS"
          value=""
        />
      </div>
    </div>
  </label>
</div>
`;

exports[`Search /w invalid 1`] = `
<div
  aria-autocomplete="list"
  aria-disabled="false"
  aria-invalid="true"
  class="Searchstyles__Box-sc-1lyvud3-0 boYLWE"
  role="combobox"
>
  <label
    class="Inputstyles__Wrapper-hc8ttv-6 iaA-dFS"
  >
    <div
      class="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
    >
      <div
        class="Inputstyles__Box-hc8ttv-1 bzkyqR"
      >
        <input
          aria-invalid="true"
          class="Inputstyles__StyledInput-hc8ttv-2 bUBKOu"
          value=""
        />
      </div>
    </div>
  </label>
</div>
`;

exports[`Search /w size 1`] = `
<div
  aria-autocomplete="list"
  aria-disabled="false"
  aria-invalid="false"
  class="Searchstyles__Box-sc-1lyvud3-0 boYLWE"
  role="combobox"
>
  <label
    class="Inputstyles__Wrapper-hc8ttv-6 gPunPw"
  >
    <div
      class="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
    >
      <div
        class="Inputstyles__Box-hc8ttv-1 bzkyqR"
      >
        <input
          aria-invalid="false"
          class="Inputstyles__StyledInput-hc8ttv-2 hbWobN"
          value=""
        />
      </div>
    </div>
  </label>
</div>
`;

exports[`Search /w size 2`] = `
<div
  aria-autocomplete="list"
  aria-disabled="false"
  aria-invalid="false"
  class="Searchstyles__Box-sc-1lyvud3-0 boYLWE"
  role="combobox"
>
  <label
    class="Inputstyles__Wrapper-hc8ttv-6 eablnU"
  >
    <div
      class="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
    >
      <div
        class="Inputstyles__Box-hc8ttv-1 bzkyqR"
      >
        <input
          aria-invalid="false"
          class="Inputstyles__StyledInput-hc8ttv-2 bUBKOu"
          value=""
        />
      </div>
    </div>
  </label>
</div>
`;

exports[`Search /w size 3`] = `
<div
  aria-autocomplete="list"
  aria-disabled="false"
  aria-invalid="false"
  class="Searchstyles__Box-sc-1lyvud3-0 boYLWE"
  role="combobox"
>
  <label
    class="Inputstyles__Wrapper-hc8ttv-6 daAZxl"
  >
    <div
      class="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
    >
      <div
        class="Inputstyles__Box-hc8ttv-1 bzkyqR"
      >
        <input
          aria-invalid="false"
          class="Inputstyles__StyledInput-hc8ttv-2 bUBKOu"
          value=""
        />
      </div>
    </div>
  </label>
</div>
`;

exports[`Search /w size 4`] = `
<div
  aria-autocomplete="list"
  aria-disabled="false"
  aria-invalid="false"
  class="Searchstyles__Box-sc-1lyvud3-0 boYLWE"
  role="combobox"
>
  <label
    class="Inputstyles__Wrapper-hc8ttv-6 dcSDdd"
  >
    <div
      class="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
    >
      <div
        class="Inputstyles__Box-hc8ttv-1 bzkyqR"
      >
        <input
          aria-invalid="false"
          class="Inputstyles__StyledInput-hc8ttv-2 QjOJw"
          value=""
        />
      </div>
    </div>
  </label>
</div>
`;

exports[`Search w/ any props 1`] = `
<div
  aria-autocomplete="list"
  aria-disabled="false"
  aria-invalid="false"
  class="Searchstyles__Box-sc-1lyvud3-0 boYLWE"
  role="combobox"
>
  <label
    class="Inputstyles__Wrapper-hc8ttv-6 daAZxl"
  >
    <div
      class="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
    >
      <div
        class="Inputstyles__Box-hc8ttv-1 bzkyqR"
      >
        <input
          aria-invalid="false"
          class="Inputstyles__StyledInput-hc8ttv-2 bUBKOu"
          title="title"
          value=""
        />
      </div>
    </div>
  </label>
</div>
`;
